File: Awful.apk/src/main/java/com/ferg/awfulapp/task/FeatureRequest.java
package com.ferg.awfulapp.task;

import android.content.Context;
import android.net.Uri;
import android.util.Log;

import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.util.AwfulError;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

/**
 * Created by matt on 8/8/13.
 */
public class FeatureRequest extends AwfulRequest<Void> {
    public FeatureRequest(Context context) {
        super(context, Constants.FUNCTION_MEMBER);
    }

    @Override
    protected String generateUrl(Uri.Builder urlBuilder) {
        return urlBuilder.appendQueryParameter(Constants.PARAM_ACTION, "accountfeatures").build().toString();
    }

    @Override
    protected Void handleResponse(Document doc) throws AwfulError {
        Element features = doc.getElementsByClass("features").first();
        boolean premium = false;
        boolean archives = false;
        boolean noads = false;
        if (features != null) {

            Elements feature_dts = features.getElementsByTag("dt");
            if (feature_dts.size() == 3) {
                premium = feature_dts.get(0).hasClass("enabled");
                archives = feature_dts.get(1).hasClass("enabled");
                noads = feature_dts.get(2).hasClass("enabled");
                try {
                    getPreferences().setBooleanPreference("has_platinum", premium);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                try {
                    getPreferences().setBooleanPreference("has_archives", archives);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                try {
                    getPreferences().setBooleanPreference("has_no_ads", noads);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        } else {
            throw new AwfulError("Feature page did not load");
        }
        Log.i("FeatureRequest", "Updated account features P:" + premium + " A:"
                + archives + " NA:" + noads);
        return null;
    }

    @Override
    protected boolean handleError(AwfulError error, Document doc) {
        return error.isCritical();
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/task/FeatureRequest.kt
package com.ferg.awfulapp.task

import android.content.Context
import android.net.Uri

import com.ferg.awfulapp.constants.Constants
import com.ferg.awfulapp.preferences.Keys
import com.ferg.awfulapp.util.AwfulError

import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import timber.log.Timber

/**
 * An AwfulRequest that fetches the features active on the user's account (platinum etc.),
 * and stores that data in AwfulPreferences.
 */
class FeatureRequest(context: Context) : AwfulRequest<Void?>(context, Constants.FUNCTION_MEMBER) {

    override fun generateUrl(urlBuilder: Uri.Builder?): String {
        return urlBuilder!!.appendQueryParameter(Constants.PARAM_ACTION, "accountfeatures").build().toString()
    }

    @Throws(AwfulError::class)
    override fun handleResponse(doc: Document): Void? {
        // grab the element containing the features info and validate its structure
        val features = doc.selectFirst(".features")?.select("dt")
        if (features == null) {
            throw AwfulError("Couldn't find features element")
        } else if (features.size != 3) {
            throw AwfulError("Unexpected number of feature elements (wanted 3, got ${features.size}")
        }

        mapOf(
                Keys.HAS_PLATINUM to features[0].enabled,
                Keys.HAS_ARCHIVES to features[1].enabled,
                Keys.HAS_NO_ADS to features[2].enabled
        ).forEach { (k, v) -> preferences.setPreference(k, v) }

        Timber.i("Updated account features\nPlatinum:${preferences.hasPlatinum} Archives:${preferences.hasArchives} NoAds:${preferences.hasNoAds}")
        return null
    }

    private val Element.enabled get() = this.hasClass("enabled")

}
